cw/util/trace_library.py

```diff
@@ -0,0 +1,231 @@
+            db.Column("bitstream", db.LargeBinary),
+            db.Column("binary", db.LargeBinary),
+            db.Column("offset", db.Integer),
+            db.Column("sample_rate", db.Integer),
+            db.Column("scope_gain", db.Float),
+            db.Column("trigger_level", db.Float),
+            db.Column("time_div", db.Float),
+            db.Column("ot_sca_commit", db.String),
+            db.Column("notes", db.String)
+        )
+        self.metadata.create_all(self.engine)
+        self.trace_mem = []
+        self.trace_mem_thr = trace_threshold
+        self.settingsDict = {}
+        self.wave_datatype = wave_datatype
+
+    def flush_to_disk(self):
+        """ Writes traces from memory into database.
+        """
+        if self.trace_mem:
+            query = db.insert(self.traces_table)
+            traces = []
+            for trace in self.trace_mem:
+                tr = asdict(trace)
+                del tr["trace_id"]
+                traces.append(tr)
+            self.session.execute(query, traces)
+            self.session.commit()
+            self.trace_mem = []
+
+    def write_to_buffer(self, trace):
+        """ Write traces into memory or into storage.
+
+        The trace first is written into memory. When reaching the trace
+        memory threshold, the traces are flushed into the database storage.
+
+        Args:
+            trace: The trace to write.
+        """
+        self.trace_mem.append(trace)
+        if (len(self.trace_mem) >= self.trace_mem_thr):
+            self.flush_to_disk()
+
+    def get_traces(self, start=0, end=0):
+        """ Get traces from database and stored into RAM.
+
+        Fetch traces from start to end from database storage into RAM.
+
+        Args:
+            start: fetch traces from trace index start
+            end: to trace index end. If no start and end is provided, all
+                 all traces are returned.
+        Returns:
+            The traces from the database.
+        """
+        self.flush_to_disk()
+
+        if end == 0:
+            query = db.select(self.traces_table)
+        else:
+            query = db.select(self.traces_table).where(
+                (self.traces_table.c.trace_id >= start) &
+                (self.traces_table.c.trace_id <= end))
+
+        ResultProxy = self.session.execute(query)
+
+        traces = []
+        for trace in ResultProxy.fetchall():
+            traces.append(Trace(**trace._mapping))
+        return traces
+
+    def get_waves(self):
+        """ Get all waves from the database.
+
+        Returns:
+            The bytearray waves from the database.
+        """
+        self.flush_to_disk()
+        traces = self.get_traces()
+        waves = []
+        for trace in traces:
+            waves.append(trace.wave)
+        return waves
+
+    def get_waves_int(self):
+        """ Get all waves from the database in the trace array format.
+
+        Returns:
+            The int waves from the database.
```

Comment:
Nit: it will return the data type defined in `wave_datatype`. So, it could also be float or double I guess. Maybe rename the functions to `get_waves()` and `get_waves_bytearray()`. In practice we will always use `get_waves()` and specify the datatype.